tensorflow_federated/python/common_libs/execution_tracing.py

```diff
@@ -0,0 +1,266 @@
+      # exit span early
+      yield None
+
+    else:
+      # Pre-fn
+      self._current_call_level += 1
+      self._log_trace('call', scope, sub_scope, fn_args, fn_kwargs)
+      yield None
+
+      # Post-fn
+      self._current_call_level -= 1
+      self._log_trace('retr', scope, sub_scope, fn_args, fn_kwargs)
+
+  @classmethod
+  def _is_traceable(cls, method_name, regex, override=False):
+    if override:
+      return True
+    match = regex.search(method_name)
+    if match is None:
+      return False
+    return True
+
+  def _log_trace(self, call_or_retr, scope, sub_scope, fn_args, fn_kwargs):
+    """Logs a function trace on call in human-readable form.
+
+    This function logs the scope and arguments with which a function has been
+    called. Nested calls via `tracing.trace` preserve depth information, so the
+    resulting log can be used to trace the AST for a given computation.
+    """
+    fn_args = fn_args or [None]
+    fn_kwargs = fn_kwargs or {}
+
+    ctx = fn_args[0]
+
+    if ctx is not None:
+      has_self = ctx.__class__.__name__ == scope
+
+      if has_self and len(fn_args) > 1:
+        fn_args = fn_args[1:]
+
+    prefix_str = self._prefix(ctx, scope)
+    main_str = '{m} {c}'.format(m=sub_scope, c=call_or_retr)
+    suffix_args = ' '.join(self._format_object(a) for a in fn_args)
+    suffix_kwargs = ' '.join(
+        self._format_object(v) for k, v in fn_kwargs.items())
+    final = '{p} {m} {sa} {sk}'.format(
+        p=prefix_str, m=main_str, sa=suffix_args, sk=suffix_kwargs)
+
+    logging.debug(final)
+
+  def _prefix(self, ctx, type_string):
+    indentation = '  ' * self._current_call_level
+    template = '{indentation}{type} {hash}'
+
+    if ctx is None:
+      return template.format(
+          indentation=indentation, type=type_string, hash='NO_ID')
+    return template.format(
+        indentation=indentation, type=type_string, hash=hash(ctx))
+
+  def _format_object(self, x):
+    strategy = self._formatting_strategies.get(
+        type(x).__name__, self._default_strategy)
+    return strategy(x)
+
+  def _format_anon_tuple(self, tup):
+    # pylint: disable=protected-access
+    names = tup._name_array
+    values = tup._element_array
+    # pylint: enable=protected-access
+    fmt_strs = []
+    for n, v in zip(names, values):
+      fmt_strs.append('{}: {}'.format(n, self._format_object(v)))
+    return '({})'.format(', '.join(fmt_strs))
+
+  def _register_formatting_strategy(self, type_string, formatting_strategy):
+    self._formatting_strategies[type_string] = formatting_strategy
+
+  # Register all types here
+  def _make_formatting_strategies(self):
+    self.register_executors()
+    self.register_executor_values()
+    self.register_computations()
+    self.register_computation_types()
+    self.register_primitives()
+
+  # pylint: disable=unnecessary-lambda,missing-function-docstring
+  def register_executors(self):
```

Comment:
It seems like all format strategies could be combined based on capabilities instead of class. So instead of:

for executor_base.Executor, EagerTensor
`'<{class} @{id}>'`

for executor_value_base.ExecutorValue
`'<{class} @{id} : {type_signature}>'`

for executor_value_base.AnonymousTuple
`'<{class} @{id} : {value}>'`

for computation_types.Type
`'<{class}>'`

for float, int
`'<{value} : {class}>'`

for None
`'-'`

for list, tuple
`'<{element 1}, {element 2}, ...>'`

for ComputationImpl ComputationPB
`'<{class} {context} @{id} : {type_signature}>'`

for IntrinsicDef
`'<{class} {uri} {type_signature} @{id}>'`

you could use: ` '<{class} {context} @{id}: {value}'`, where:

* special case None
* everything has a `class`
* ComputationImpl and ComputationPB have a `context` based on proto
* IntrinsicDef has a `context` based on URI
* everything else has no `context`
* everything as an `id`
* things with type_signature have a `value` of type_signature
* things with a value have a `value` of value
* everything else has no `value`

This make the problem a lot smaller... And makes the representations more consistent.

Maybe we can use duck-typing to get rid of the dependencies? Or maybe drop those differences? Or maybe fork on base classes instead of the individual subclasses? WDYT?

Reply:
This seems useful! I do think the differences matter to make the end result readable, so I'm inclined to choose duck-typing or forking. I think I slightly prefer duck-typing; we can think of these differences as "traits" that are agnostic to the types, which means we can avoid the imports. The only improvement I can imagine would be to let the types themselves handle this, or perhaps to pass in as trace_opts, but both of those alternatives seem invasive for something that's meant to be a simple & lightweight developer tool.

If there are no objections, I'll refactor the PR with all this in mind in the next day or two.

Reply:
+1 to "let the types themselves handle this explicitly". This would be nice, and I agree this is more invasive.